osquery/events/darwin/endpointsecurity.cpp

```diff
@@ -0,0 +1,232 @@
+  return p->signing_id.length > 0 ? p->signing_id.data : "";
+}
+
+static inline std::string getTeamId(const es_process_t* p) {
+  return p->team_id.length > 0 ? p->team_id.data : "";
+}
+
+static inline std::string getStringFromToken(es_string_token_t* t) {
+  return t->length > 0 && t->data != nullptr ? t->data : "";
+}
+
+static inline std::string getCwdPathFromPid(pid_t pid) {
+  struct proc_vnodepathinfo vpi {};
+  auto bytes = proc_pidinfo(pid, PROC_PIDVNODEPATHINFO, 0, &vpi, sizeof(vpi));
+  return bytes <= 0 ? "" : vpi.pvi_cdir.vip_path;
+}
+
+static inline std::string getCDHash(const es_process_t* p) {
+  std::stringstream hash;
+  for (unsigned char i : p->cdhash) {
+    hash << std::hex << std::setfill('0') << std::setw(2)
+         << static_cast<unsigned int>(i);
+  }
+  auto s = hash.str();
+  return s.find_first_not_of(s.front()) == std::string::npos ? "" : s;
+}
+
+static inline void getProperties(const es_process_t* p,
+                                 const EndpointSecurityEventContextRef& ec) {
+  auto audit_token = p->audit_token;
+  ec->pid = audit_token_to_pid(audit_token);
+  ec->parent = p->ppid;
+  ec->original_parent = p->original_ppid;
+
+  ec->path = getPath(p);
+  ec->cwd = getCwdPathFromPid(ec->pid);
+
+  ec->uid = audit_token_to_ruid(audit_token);
+  ec->euid = audit_token_to_egid(audit_token);
+  ec->gid = audit_token_to_rgid(audit_token);
+  ec->egid = audit_token_to_egid(audit_token);
+
+  ec->signing_id = getSigningId(p);
+  ec->team_id = getTeamId(p);
+  ec->cdhash = getCDHash(p);
+  ec->platform_binary = p->is_platform_binary;
+
+  auto user = getpwuid(ec->uid);
+  ec->username = std::string(user->pw_name);
+
+  ec->cwd = getCwdPathFromPid(ec->pid);
+}
+
+void EndpointSecurityPublisher::handleMessage(const es_message_t* message) {
+  if (message == nullptr) {
+    return;
+  }
+
+  if (message->action_type == ES_ACTION_TYPE_AUTH) {
+    return;
+  }
+
+  auto ec = createEventContext();
+
+  ec->version = message->version;
+  if (ec->version >= 2) {
+    ec->seq_num = message->seq_num;
+  }
+
+  if (ec->version >= 4) {
+    ec->global_seq_num = message->global_seq_num;
+  }
+
+  getProperties(message->process, ec);
+
+  switch (message->event_type) {
+  case ES_EVENT_TYPE_NOTIFY_EXEC: {
+    ec->es_event = ES_EVENT_TYPE_NOTIFY_EXEC;
+    ec->event_type = "exec";
+
+    getProperties(message->event.exec.target, ec);
+    ec->argc = es_exec_arg_count(&message->event.exec);
+    std::stringstream args;
+    for (auto i = 0; i < ec->argc; i++) {
+      auto arg = es_exec_arg(&message->event.exec, i);
+      auto s = getStringFromToken(&arg);
+      args << s << ' ';
+    }
+    ec->args = args.str();
+
+    ec->envc = es_exec_env_count(&message->event.exec);
+    std::stringstream envs;
+    for (auto i = 0; i < ec->envc; i++) {
+      auto env = es_exec_env(&message->event.exec, i);
+      auto s = getStringFromToken(&env);
+      envs << s << ' ';
+    }
+    ec->envs = envs.str();
```

Comment:
Ditto from above, considering moving the content or keeping the string stream in a temporary scope.